ntbs-service/DataMigration/ImportValidator.cs

```diff
@@ -89,48 +128,54 @@ public ImportValidator(IImportLogger logger, IReferenceDataRepository referenceD
-                                   || Attribute.IsDefined(prop, typeof(MaxLengthAttribute))))
-                .ForEach(prop => prop.SetValue(entity, TrimAndCleanString((string)prop.GetValue(entity))));
+            var messages = new List<string>();
+            var lostData = new List<string>();
 
-            // Clean strings properties of this object's entity children
-            entity?
-                .GetType()
-                .GetProperties()
-                .Where(prop => Attribute.IsDefined(prop, typeof(ValidationChildAttribute)))
-                .ForEach(prop => TrimAndCleanStringProperties(prop.GetValue(entity)));
+            // The order we clean properties is important, because if we overwrite one value (eg. AdultsScreened) with
+            // null then we also need to overwrite the values that depend on it (eg. AdultsActiveTB)
+            var propertyNames = new List<string>
+            {
+                nameof(ContactTracing.AdultsScreened), nameof(ContactTracing.ChildrenScreened),
+                nameof(ContactTracing.AdultsActiveTB), nameof(ContactTracing.ChildrenActiveTB),
+                nameof(ContactTracing.AdultsLatentTB), nameof(ContactTracing.ChildrenLatentTB),
+                nameof(ContactTracing.AdultsStartedTreatment), nameof(ContactTracing.ChildrenStartedTreatment),
+                nameof(ContactTracing.AdultsFinishedTreatment), nameof(ContactTracing.ChildrenFinishedTreatment),
+            };
 
-            // Clean strings properties of this object's enumerable children
-            entity?
-                .GetType()
-                .GetProperties()
-                .Where(prop => Attribute.IsDefined(prop, typeof(ValidationChildEnumerableAttribute)))
-                .Select(prop => prop.GetValue(entity))
-                .Where(enumerable => enumerable != null)
-                .OfType<IEnumerable<object>>()
-                // Flatten
-                .SelectMany(element => element)
-                .ForEach(TrimAndCleanStringProperties);
-        }
+            foreach (var propertyName in propertyNames)
+            {
+                var property = notification.ContactTracing
+                    .GetType()
+                    .GetProperty(propertyName);
+                var propertyValue = property?.GetValue(notification.ContactTracing);
 
-        private static string TrimAndCleanString(string s) => s?.Trim().Replace("\t", " ");
+                var validationResults =
+                    ValidateProperty(notification.ContactTracing, propertyValue, propertyName).ToList();
+                if (validationResults.Any())
+                {
+                    property?.SetValue(notification.ContactTracing, null);
+                    messages.AddRange(validationResults.Select(result => result.ErrorMessage));
+                    lostData.Add($"{propertyName}: {propertyValue}");
+                }
```

Comment:
I would probably extract the content of the `foreach` loop into a separate method, because it feels like a lower level of abstraction than the rest of this method.

Reply:
Yeah sure. I did consider this but was a bit unsure on how to deal with returning the values that need to go into `messages` and `lostData`. I've gone with passing references to those lists into the new function, but this does feel a little weird. I could return a tuple of the bits of data, but then there's more processing to do where I'm running the function. What do you think?

Reply:
I probably would have gone with returning a tuple (or even have a private class encapsulating the output of the method, as that may be clearer). In my opinion it's slightly cleaner than passing references, but I'm happy enough with what you've got.